package vetclinic.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import vetclinic.auth.JwtAuthenticationFilter;
import vetclinic.auth.JwtTokenProvider;
import vetclinic.repository.UserSessionRepository;

@Configuration
public class SecurityConfig {

    private final JwtTokenProvider jwt;
    private final UserSessionRepository sessionRepo;

    public SecurityConfig(JwtTokenProvider jwt,
                          UserSessionRepository sessionRepo) {
        this.jwt = jwt;
        this.sessionRepo = sessionRepo;
    }

    @Bean
    public JwtAuthenticationFilter jwtAuthenticationFilter() {
        return new JwtAuthenticationFilter(jwt, sessionRepo);
    }

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        http
                .csrf(csrf -> csrf.disable())
                .authorizeHttpRequests(auth -> auth

                        // открытые эндпоинты
                        .requestMatchers("/auth/**").permitAll()

                        // только ADMIN может менять данные
                        .requestMatchers(org.springframework.http.HttpMethod.POST, "/api/**").hasRole("ADMIN")
                        .requestMatchers(org.springframework.http.HttpMethod.PUT, "/api/**").hasRole("ADMIN")
                        .requestMatchers(org.springframework.http.HttpMethod.DELETE, "/api/**").hasRole("ADMIN")

                        // все авторизованные могут читать
                        .requestMatchers(org.springframework.http.HttpMethod.GET, "/api/**").authenticated()

                        .anyRequest().authenticated()
                )
                .addFilterBefore(
                        jwtAuthenticationFilter(),
                        UsernamePasswordAuthenticationFilter.class
                )
                .formLogin(form -> form.disable())
                .httpBasic(basic -> {});

        return http.build();
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }
}
